modules/build/src/main/scala/scala/build/input/Inputs.scala

```diff
@@ -302,19 +313,30 @@ object Inputs {
       else if isRunWithShebang && os.exists(path) then
         if isShebangScript(String(content)) then Right(Seq(Script(dir, subPath)))
         else
-          Left(s"""$arg does not contain shebang header
-                  |possible fixes:
-                  |  Add '#!/usr/bin/env scala-cli shebang' to the top of the file
-                  |  Add extension to the file's name e.q. '.sc'
-                  |""".stripMargin)
+          Left(if isShebangCapableShell then
+            s"""$unrecognizedSourceError,
+               |to use a script with no file extensions add shebang header pointing to '$fullProgramCall' to the top of the file
+               |""".stripMargin
+          else unrecognizedSourceError)
       else {
         val msg =
           if os.exists(path) then
-            if isShebangScript(String(content)) then
-              s"$arg scripts with no file extension should be run with 'scala-cli shebang'"
-            else
-              s"$arg: unrecognized source type (expected .scala or .sc extension, or a directory)"
-          else s"$arg: not found"
+            if isShebangCapableShell && isShebangScript(String(content)) then
```

Comment:
NIT I feel a match clause on `ScalaCliInvokeData` could improve readability of this overly complex nest of if clauses...